src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -1542,9 +1577,11 @@ void Terminal::_updateUrlDetection()
     }
 }
 
+// NOTE: This is the version of AddMark that comes from the UI. The VT api call into this too.
 void Terminal::AddMark(const Microsoft::Console::VirtualTerminal::DispatchTypes::ScrollMark& mark,
                        const til::point& start,
-                       const til::point& end)
+                       const til::point& end,
+                       const bool fromUi)

```

Comment:
I believe that using indices would be better than using `.back()` and adding `.insert()`.
Rust exposes this problem much more obviously, as you can't "borrow" foreign data easily there without introducing lifetime issues. A common solution is to instead save a numeric index (or similar) of where the data is stored at. In this case you could store a `size_t _currentPromptMarkIndex` which stores the index of the mark inside `_scrollMarks`, so that you can safely retrieve it from `_scrollMarks` at a later time.
It's basically exactly like the `_currentPrompt` pointer since it's still sort of a (relative) memory location, but unlike it, it's much easier to prove that it doesn't reference invalid data (by checking that `_currentPromptMarkIndex < .size()`). This way you don't need to `insert()` user marks at the beginning either, allowing you to remove this boolean parameter. Such a numeric index doesn't fix any "stateful" bugs of course, but it does prevent memory lifetime issues, as was the case with the raw pointer.